pyccel/parser/semantic.py

```diff
@@ -487,10 +487,17 @@ def _garbage_collector(self, expr):
         """
         Search in a CodeBlock if no trailing Return Node is present add the needed frees.
         """
+
+        deallocs = []
         if len(expr.body)>0 and not isinstance(expr.body[-1], Return):
-            deallocs = [Deallocate(i) for i in self._allocs[-1]]
-        else:
-            deallocs = []
+            for i in self._allocs[-1]:
+                if isinstance(i, DottedVariable):
+                    if isinstance(i.lhs.dtype, CustomDataType) and not self._current_function == '__Pyccel__del__':
+                        continue
+                if isinstance(i, ConstructorCall):
+                    deallocs.append(Deallocate(i.cls_variable))
+                    continue
+                deallocs.append(Deallocate(i))
```

Comment:
I find this more readable:
```suggestion
                else:
                    deallocs.append(Deallocate(i))
```

Reply:
Fixed https://github.com/pyccel/pyccel/pull/1508/commits/4464522edd12b45c32f6d7ee46ababb878edcd2f.